### DISCUSSION TITLE
`brew shellenv` is generating no output

### DISCUSSION THREAD

**Discussion by luckman212** [Homebrew/discussions#2547] [Everyday usage] [answered]
_(again, sorry to post here, but [the issue tracker is still down](https://github.com/Homebrew/discussions/discussions/2505))_

At some recent point, `brew shellenv` stopped producing any output on my system. It used to work. I figured out that adding `-v` produces the expected output:

```
$ brew shellenv
(nothing)
$ brew -v shellenv
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
export HOMEBREW_SHELLENV_PREFIX="/opt/homebrew";
export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";

$ brew update
Already up-to-date.

$ brew doctor
Your system is ready to brew.

$ brew config
HOMEBREW_VERSION: 3.3.5-61-g4c3f325
ORIGIN: https://github.com/Homebrew/brew
HEAD: 4c3f3253473e357c9291e6f10df856ee7019d7a4
Last commit: 15 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 0d8fc4b6b88627d63f7d78207cc69d476794c1a4
Core tap last commit: 38 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_EDITOR: /usr/local/bin/subl -w
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 13.0.0 build 1300
Git: 2.34.1 => /opt/homebrew/bin/git
Curl: 7.77.0 => /usr/bin/curl
macOS: 12.0.1-arm64
CLT: 13.1.0.0.1.1633545042
Xcode: 13.1
Rosetta 2: false
```

**Comment by Bo98** [comment, marked as answer]
Is the `brew shellenv` variables already exported? It won't output anything if so. Check `echo $HOMEBREW_SHELLENV_PREFIX`.

**Reply by luckman212** [reply]
Yes, the variables are already exported. Is this a change? It wasn't the previous behavior as far as I remember.

**Reply by Bo98** [reply]
This was changed to fix https://github.com/Homebrew/install/issues/559.

**Reply by luckman212** [reply]
Ah ok. TIL. It might be nice if a message was logged to `stderr` if this path is hit. Something like an `echo 1>&2 "env variables already exported, run with -v to override"`. This should not break anything.

**Comment by carlocab** [comment]
I've opened Homebrew/brew#12495 to adjust this behaviour.

**Comment by wheelerlaw** [comment]
`brew -v shellenv` no longer prints the output. Was that option removed? I'd like to know what `shellenv` is printing because I don't have it memorized off the top of my head.

**Reply by gromgit** [reply]
Explanatory comment in the `shellenv` script:
```
To help guarantee idempotence, this command produces no output when Homebrew's `bin` and `sbin` directories are first and second
respectively in your `PATH`.
```
So if your `PATH` starts with the Homebrew `bin` and `sbin` directories, you can expect no output.

> I'd like to know what `shellenv` is printing because I don't have it memorized off the top of my head.

Per the above comment, you can force output by prepending something to your `PATH`:
```bash
PATH=/dummy:$PATH brew shellenv
```

**Reply by wheelerlaw** [reply]
Right, I expected just `brew shellenv` to not produce anything, but I saw in the original discussion post that they used `-v` to force the ourput. I tried it out myself and noticed that it didn't work, so I was wondering if at some point it had disappeared.

**Reply by gromgit** [reply]
As far as I know:
1. `brew` will move any `-v` flag that's first in its arguments to the last, so `brew -v shellenv` is actually evaluated as `brew shellenv -v`.
2. The `shellenv` subcommand doesn't take any arguments. The old (Ruby-based?) implementation might have done something with it, but that was replaced by the current Bash implementation in **2018**.

Out of curiosity, when was the last time that `brew -v shellenv` worked for you?

**Reply by wheelerlaw** [reply]
I had never tried `brew -v shellenv` until I saw this discussion several days ago.

**Comment by warsaw** [comment]
I just ran into this problem and nothing I do, nor any of the suggestions above, get `brew shellenv` to print anything.  I'm on Homebrew 4.5.2, macOS 15.5, Apple Silicon.  The biggest problem for me was that `$MANPATH` wasn't getting set, so I just added this manually to my dot files:

```
export MANPATH="$($brew --prefix)/share/man${MANPATH+:$MANPATH}:";
```

(I set `$brew` elsewhere.)

**Reply by luckman212** [reply]
Not experiencing that on my system. I'm also on 15.5 (arm) and Brew 4.5.2. Both `brew shellenv` and `brew shellenv -v` produce the same output:

```
$ brew shellenv
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
PATH="/opt/homebrew/bin:/opt/homebrew/sbin:/Users/luke/.n/bin:/Users/luke/.pipx/bin:/Users/luke/.virtualenvs/bin:/Users/luke/.python/bin:/Users/luke/.orbstack/bin:/Users/luke/.cargo/bin:/Users/luke/go/bin:/Users/luke/Sync/Scripts:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/Applications/Wireshark.app/Contents/MacOS:/Applications/iTerm.app/Contents/Resources/utilities"; export PATH;
[ -z "${MANPATH-}" ] || export MANPATH=":${MANPATH#:}";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```
What's your `brew --config` look like?

**Reply by gromgit** [reply]
> nothing I do, nor any of the suggestions above, get `brew shellenv` to print anything.

As I mentioned in [another comment](https://github.com/orgs/Homebrew/discussions/2547#discussioncomment-4032757), the only test that `brew shellenv` does to decide whether to show _any_ output is If your PATH already starts with `${HOMEBREW_PREFIX}/bin:${HOMEBREW_PREFIX}/sbin`. If you're already setting that manually before running `brew shellenv`, then no output will be shown.

What's the output of `PATH=/dummy:$PATH brew shellenv`?

> The biggest problem for me was that `$MANPATH` wasn't getting set

That's not actually a problem in pretty much all modern *nixes (including macOS), as the widely-used `man` implementation defaults its search path based on the path that the command being `man`'d is found. This is why `MANPATH= man gfind` works even when it shouldn't, because `man` looks for `gfind` and finds it in `${HOMEBREW_PREFIX}/bin`, so it automatically looks in `${HOMEBREW_PREFIX}/man` and `${HOMEBREW_PREFIX}/share/man`. You can `man manpath` to learn more about the default man search path.

If you have a non-empty `MANPATH`, you can inject the default man search path anywhere you want:
1. Before all other directories: add a colon at the beginning (`MANPATH=:${MANPATH}`)
2. After all other directories: add a colon at the end (`MANPATH=${MANPATH}:`)
3. Somewhere in the middle: add _two_ colons where you want to insert the default search path (`MANPATH=/blah:/foo::/baz:/burp`)

This line @luckman212 quoted from `brew shellenv`:
>```
>[ -z "${MANPATH-}" ] || export MANPATH=":${MANPATH#:}";
>```

implements option 1 _and_ correctly does nothing with an empty `MANPATH`.

**Reply by warsaw** [reply]
Thanks for the explanation about `$MANPATH`.  I think with that -- and a little testing -- it's clear that `brew shellenv` is working for me as intended.

One suggestion I'd make is for `brew shellenv -d` or `-v` to actual print out what it _would_ do if not for the pre-existing `$PATH` setting.  I think a debug mode would help clear up these confusions.

**Comment by warsaw** [comment]
> What's your brew --config look like?

```
HOMEBREW_VERSION: 4.5.2
ORIGIN: https://github.com/Homebrew/brew
HEAD: 6f39076b3c2251994419215279d0525ef667fc31
Last commit: 13 days ago
Branch: stable
Core tap JSON: 21 May 19:23 UTC
Core cask tap JSON: 21 May 19:23 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: emacsclient
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_NO_ANALYTICS: set
HOMEBREW_NO_ENV_HINTS: set
Homebrew Ruby: 3.4.3 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.4.3/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 17.0.0 build 1700
Git: 2.49.0 => /opt/homebrew/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.5-arm64
CLT: 16.3.0.0.1.1742442376
Xcode: 16.3
Rosetta 2: false
```

Could it have something to do with `HOMEBREW_NO_ENV_HINTS`?

**Comment by jhadida** [comment]
I don't know if this will help anyone else, but in my case setting up Xcode with 
```
sudo xcode-select --install
```
appears to have been needed for `brew shellenv` to output something.
